File: software/mosek/7/tools/examples/python/sdo1.py
##
#   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.
#
#   File:      sdo1.py
#
#   Purpose:   Demonstrates how to solve a small mixed semidefinite and conic quadratic
#              optimization problem using the MOSEK Python API.
##

import sys
import mosek

# If numpy is installed, use that, otherwise use the 
# Mosek's array module.
try:
    from numpy import array,zeros,ones
except ImportError:
    from mosek.array import array, zeros, ones
    
# Since the value of infinity is ignored, we define it solely
# for symbolic purposes
inf = 0.0

# Define a stream printer to grab output from MOSEK
def streamprinter(text):
    sys.stdout.write(text)
    sys.stdout.flush()
    
def main ():
  # Make mosek environment
  env = mosek.Env()
  
  # Create a task object and attach log stream printer
  task = env.Task(0,0)
  task.set_Stream(mosek.streamtype.log, streamprinter)

  # Bound keys for constraints
  bkc = [mosek.boundkey.fx,
         mosek.boundkey.fx]

  # Bound values for constraints
  blc = [1.0, 0.5]
  buc = [1.0, 0.5]
  
  # Below is the sparse representation of the A
  # matrix stored by row. 
  asub = [ array([0]),
           array([1, 2])]
  aval = [ array([1.0]),
           array([1.0, 1.0]) ]

  conesub = [0, 1, 2]

  barci = [0, 1, 1, 2, 2]
  barcj = [0, 0, 1, 1, 2]
  barcval = [2.0, 1.0, 2.0, 1.0, 2.0]

  barai   = [array([0, 1, 2]), 
             array([0, 1, 2, 1, 2, 2])]
  baraj   = [array([0, 1, 2]), 
             array([0, 0, 0, 1, 1, 2])]
  baraval = [array([1.0, 1.0, 1.0]), 
             array([1.0, 1.0, 1.0, 1.0, 1.0, 1.0])]

  numvar = 3
  numcon = len(bkc)
  BARVARDIM = [3]

  # Append 'numvar' variables.
  # The variables will initially be fixed at zero (x=0). 
  task.appendvars(numvar)

  # Append 'numcon' empty constraints.
  # The constraints will initially have no bounds. 
  task.appendcons(numcon)
  
  # Append matrix variables of sizes in 'BARVARDIM'.
  # The variables will initially be fixed at zero. 
  task.appendbarvars(BARVARDIM)

  # Set the linear term c_0 in the objective.
  task.putcj(0, 1.0)
  
  for j in range(numvar):
    # Set the bounds on variable j
    # blx[j] <= x_j <= bux[j] 
    task.putvarbound(j, mosek.boundkey.fr, -inf, +inf)
  
  for i in range(numcon):
    # Set the bounds on constraints.
    # blc[i] <= constraint_i <= buc[i]
    task.putconbound(i, bkc[i], blc[i], buc[i])
    
    # Input row i of A 
    task.putarow(i,                  # Constraint (row) index.
                 asub[i],            # Column index of non-zeros in constraint j.
                 aval[i])            # Non-zero values of row j. 
  
  task.appendcone(mosek.conetype.quad,
                  0.0,
                  conesub)

  symc = \
    task.appendsparsesymmat(BARVARDIM[0], 
                            barci, 
                            barcj, 
                            barcval)

  syma0 = \
    task.appendsparsesymmat(BARVARDIM[0], 
                            barai[0], 
                            baraj[0], 
                            baraval[0])
    
  syma1 = \
    task.appendsparsesymmat(BARVARDIM[0], 
                            barai[1], 
                            baraj[1], 
                            baraval[1])

  task.putbarcj(0, [symc], [1.0])

  task.putbaraij(0, 0, [syma0], [1.0])
  task.putbaraij(1, 0, [syma1], [1.0])

  # Input the objective sense (minimize/maximize)
  task.putobjsense(mosek.objsense.minimize)

  task.writedata("sdo1.task")

  # Solve the problem and print summary
  task.optimize()
  task.solutionsummary(mosek.streamtype.msg)
  
  # Get status information about the solution
  prosta = task.getprosta(mosek.soltype.itr)
  solsta = task.getsolsta(mosek.soltype.itr)

  if (solsta == mosek.solsta.optimal or 
      solsta == mosek.solsta.near_optimal):
    xx = zeros(numvar, float)
    task.getxx(mosek.soltype.itr, xx)

    lenbarvar = BARVARDIM[0] * (BARVARDIM[0]+1) / 2
    barx = zeros(int(lenbarvar), float)
    task.getbarxj(mosek.soltype.itr, 0, barx)

    print("Optimal solution:\nx=%s\nbarx=%s" % (xx,barx))
  elif (solsta == mosek.solsta.dual_infeas_cer or
        solsta == mosek.solsta.prim_infeas_cer or
        solsta == mosek.solsta.near_dual_infeas_cer or
        solsta == mosek.solsta.near_prim_infeas_cer):
    print("Primal or dual infeasibility certificate found.\n")
  elif solsta == mosek.solsta.unknown:
    print("Unknown solution status")
  else:
    print("Other solution status")

# call the main function
try:
    main ()
except mosek.Exception as e:
    print ("ERROR: %s" % str(e.errno))
    if e.msg is not None:
        print ("\t%s" % e.msg)
        sys.exit(1)
except:
    import traceback
    traceback.print_exc()
    sys.exit(1)
sys.exit(0)

File: software/mosek/7/tools/examples/java/sdo1.java
package com.mosek.example;
import mosek.Env;
/* 
   Copyright: Copyright (c) MOSEK ApS, Denmark. All rights reserved.

   File:      sdo1.java

   Purpose:   Solves the following small semidefinite optimization problem 
              using the MOSEK API.

     minimize    Tr [2, 1, 0; 1, 2, 1; 0, 1, 2]*X + x0
 
     subject to  Tr [1, 0, 0; 0, 1, 0; 0, 0, 1]*X + x0           = 1
                 Tr [1, 1, 1; 1, 1, 1; 1, 1, 1]*X      + x1 + x2 = 0.5
                 (x0,x1,x2) \in Q,  X \in PSD
*/

public class sdo1 
{

  public static void main(String[] argv)
  {
    mosek.Env  env  = null;
    mosek.Task task = null;


    int    numcon      = 2;  /* Number of constraints.              */
    int    numvar      = 3;  /* Number of conic quadratic variables */
    int    numanz      = 3;  /* Number of non-zeros in A            */
    int    numbarvar   = 1;  /* Number of semidefinite variables    */
    int    dimbarvar[] = {3};         /* Dimension of semidefinite cone */
    int    lenbarvar[] = {3*(3+1)/2}; /* Number of scalar SD variables  */

    mosek.Env.boundkey bkc[] = { mosek.Env.boundkey.fx,
                                 mosek.Env.boundkey.fx };
    double[]     blc     = { 1.0, 0.5 };
    double[]     buc     = { 1.0, 0.5 };

    int[]        barc_i  = {0, 1, 1, 2, 2},
                 barc_j  = {0, 0, 1, 1, 2};
    double[]     barc_v  = {2.0, 1.0, 2.0, 1.0, 2.0};

    int[][]      asub    = {{0}, {1, 2}}; /* column subscripts of A */
    double[][]   aval    = {{1.0}, {1.0, 1.0}};

    int[][]      bara_i  = { {0,   1,   2},   {0,   1 ,  2,   1,   2,   2 } },
                 bara_j  = { {0,   1,   2},   {0,   0 ,  0,   1,   1,   2 } };
    double[][]   bara_v  = { {1.0, 1.0, 1.0}, {1.0, 1.0, 1.0, 1.0, 1.0, 1.0}};
    int[]        conesub = { 0, 1, 2};


    try
    {
      // Make mosek environment. 
      env  = new mosek.Env ();
      // Create a task object. 
      task = new mosek.Task (env, 0, 0);

      // Directs the log task stream to the user specified
      // method task_msg_obj.stream
      task.set_Stream(
        mosek.Env.streamtype.log,
        new mosek.Stream() 
          { public void stream(String msg) { System.out.print(msg); }});

      /* Append 'NUMCON' empty constraints.
         The constraints will initially have no bounds. */
      task.appendcons(numcon);

      /* Append 'NUMVAR' variables.
         The variables will initially be fixed at zero (x=0). */
      task.appendvars(numvar);

      /* Append 'NUMBARVAR' semidefinite variables. */
      task.appendbarvars(dimbarvar);

      /* Optionally add a constant term to the objective. */
      task.putcfix(0.0);

      /* Set the linear term c_j in the objective.*/
      task.putcj(0,1.0);

      for (int j = 0; j<numvar; ++j)
        task.putvarbound(j,mosek.Env.boundkey.fr,-0.0,0.0);

      /* Set the linear term barc_j in the objective.*/  
      {
        long[] idx = new long[1];
        double[] falpha = { 1.0 };
        idx[0] = task.appendsparsesymmat(dimbarvar[0],
                                         barc_i,
                                         barc_j,
                                         barc_v);
        task.putbarcj(0, idx, falpha); 
      }

      /* Set the bounds on constraints.
        for i=1, ...,numcon : blc[i] <= constraint i <= buc[i] */

      for(int i = 0; i < numcon; ++i)
        task.putconbound(i,           /* Index of constraint.*/
                         bkc[i],      /* Bound key.*/
                         blc[i],      /* Numerical value of lower bound.*/
                         buc[i]);     /* Numerical value of upper bound.*/

      /* Input A row by row */
      for (int i=0; i<numcon; ++i)
        task.putarow(i,
                     asub[i],
                     aval[i]);

      /* Append the conic quadratic cone */
      task.appendcone(mosek.Env.conetype.quad,
                      0.0,
                      conesub);

      /* Add the first row of barA */
      {
        long[] idx = new long[1];
        double[] falpha = {1.0};
        task.appendsparsesymmat(dimbarvar[0],
                                bara_i[0],
                                bara_j[0],
                                bara_v[0],
                                idx);

        task.putbaraij(0, 0, idx, falpha);
      }

      {
        long[] idx = new long[1];
        double[] falpha = {1.0};
        /* Add the second row of barA */
        task.appendsparsesymmat(dimbarvar[0],
                                bara_i[1],
                                bara_j[1],
                                bara_v[1],
                                idx);

        task.putbaraij(1, 0, idx, falpha);
      }

      /* Run optimizer */
      task.optimize();

      /* Print a summary containing information
         about the solution for debugging purposes*/
      task.solutionsummary (mosek.Env.streamtype.msg);

      mosek.Env.solsta[] solsta = new mosek.Env.solsta[1];
      task.getsolsta (mosek.Env.soltype.itr,solsta);

      switch(solsta[0])
      {
      case optimal:
      case near_optimal:
        double[] xx = new double[numvar];
        double[] barx = new double[lenbarvar[0]];

        task.getxx(mosek.Env.soltype.itr,xx);
        task.getbarxj(mosek.Env.soltype.itr,    /* Request the interior solution. */
                      0,
                      barx);
        System.out.println("Optimal primal solution");
        for(int i=0; i < numvar; ++i)
          System.out.println("x[" + i + "]   : " + xx[i]);

        for(int i=0; i<lenbarvar[0]; ++i)
          System.out.println("barx[" + i + "]: " + barx[i]);
        break;
      case dual_infeas_cer:
      case prim_infeas_cer:
      case near_dual_infeas_cer:
      case near_prim_infeas_cer:  
        System.out.println("Primal or dual infeasibility certificate found.");
        break;
      case unknown:
        System.out.println("The status of the solution could not be determined.");
        break;  
      default:
        System.out.println("Other solution status.");
        break;
      }
    }
    finally
    {
      if (task != null) task.dispose();
      if (env != null)  env.dispose();
    }
  }
}

